test/integration/client-side-encryption/client_side_encryption.prose.test.js

```diff
@@ -44,6 +61,437 @@ describe('Client Side Encryption Prose Tests', function () {
     'base64'
   );
 
+  context('when kmip is the kms provider', metadata, async function () {
+    const autoEncryptionOptions = {
+      keyVaultNamespace,
+      kmsProviders: {
+        kmip: {
+          endpoint: 'localhost:5698'
+        }
+      },
+      tlsOptions: {
+        kmip: {
+          tlsCAFile: process.env.KMIP_TLS_CA_FILE,
+          tlsCertificateKeyFile: process.env.KMIP_TLS_CERT_FILE
+        }
+      }
+    };
+    let client;
+    let clientEncryption;
+    let clientEncryptionInvalid;
+
+    before(async function () {
+      client = this.configuration.newClient();
+      if (process.env.LOAD_BALANCER) this.skip();
+      await client.connect();
+      const mongodbClientEncryption = this.configuration.mongodbClientEncryption;
+      clientEncryption = new mongodbClientEncryption.ClientEncryption(client, {
+        ...autoEncryptionOptions,
+        bson: BSON
+      });
+      clientEncryptionInvalid = new mongodbClientEncryption.ClientEncryption(client, {
+        ...autoEncryptionOptions,
+        bson: BSON,
+        kmsProviders: {
+          kmip: {
+            endpoint: 'invalid.localhost:5698'
+          }
+        }
+      });
+      await dropCollection(client.db(keyVaultDbName), keyVaultCollName);
+      await dropCollection(client.db(keyVaultDbName), keyVaultCollName);
+    });
+
+    after(async function () {
+      await client.close();
+    });
+
+    context('when encrypting with kmip', metadata, async function () {
+      context('when not providing an endpoint in the master key', metadata, async function () {
```

Comment:
Can we identify the prose tests a little bit better throughout? The comment blocks for 10-12 help, but I think including something like `Test case #10: <description>` in the describe block would be even more clear. Then there are also tests below that are more difficult to tie precisely to the prose cases in the spec. If we had the outer blocks correspond to the Prose test section headings (e.g. "Custom Endpoint Test" > "kmip" (if we want to group) > "Test case #x"; with whatever extra blocks are needed in between for setup), it would be easier to audit later.

Reply:
related, I was having trouble finding the origin, the phrasing of the test names over in the specs might not work exactly right for us inside our `it` blocks but let's copy paste the names into the comments so they're searchable

Reply:
I've updated the test descriptions to match the names in the spec prose test definitions now, and moved the KMIP custom endpoint tests in with the other custom endpoint tests so they follow the same pattern and order.